daemon/algod/api/server/v2/handlers.go

```diff
@@ -1153,41 +1153,30 @@ func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64,
 	castedMax := nilToZero(params.Max)
 	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
 
-	// compute `maxReturnNum` from `castedMax` and `maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication`
-	// `maxReturnNum == 0` for returning everything
-	// - `castedMax == 0` and `maxBoxThreshold == 0`,
-	//       return everything
-	// - `castedMax == 0` and `maxBoxThreshold > 0`,
-	//       returns up to `maxBoxThreshold`, check if exceeds
-	// - `castedMax > 0` and `maxBoxThreshold == 0`,
-	//       return up to `castedMax`
-	// - `castedMax > 0` and `maxBoxThreshold > 0`,
-	//       return up to min(castedMax, maxBoxThreshold),
-	//       if return size is `maxBoxThreshold`, check if exceeds
-
-	var boxKeys []string
-	var err error
-
-	dominatedByQryParams := castedMax > 0 && (maxBoxThreshold >= castedMax || maxBoxThreshold == 0)
-	returnsAll := castedMax == 0 && maxBoxThreshold == 0
-
-	if dominatedByQryParams || returnsAll {
-		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, castedMax)
-		if err != nil {
-			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
-		}
-	} else {
-		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, maxBoxThreshold+1)
-		if err != nil {
-			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
-		}
-		if uint64(len(boxKeys)) > maxBoxThreshold {
-			v2.Log.Info("MaxAPIBoxPerApplication limit %d exceeded", maxBoxThreshold)
-			return ctx.JSON(http.StatusBadRequest, generated.ErrorResponse{
-				Message: "Result limit exceeded",
-				Data:    nil,
-			})
+	algodSupportsUnlimitedResults := maxBoxThreshold == 0
+
+	maxKeys := func() uint64 {
+		noRequestProvidedLimit := castedMax == 0
+		dominatedByQryParams := castedMax > 0 && (maxBoxThreshold >= castedMax || algodSupportsUnlimitedResults)
+		returnsAll := noRequestProvidedLimit && algodSupportsUnlimitedResults
+
+		if dominatedByQryParams || returnsAll {
+			return castedMax
 		}
+		return maxBoxThreshold + 1
+	}
+
+	boxKeys, err := ledger.LookupKeysByPrefix(lastRound, keyPrefix, maxKeys())
+	if err != nil {
+		return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+	}
+
+	if !algodSupportsUnlimitedResults && uint64(len(boxKeys)) > maxBoxThreshold {
```

Comment:
I still can't see how this could ever happen under normal circumstances. I understand that `maxBoxThreshold + 1` is sometimes provided and then one could go over the limit. I guess this is a way to tell that there are more boxes to see than the server can make available.... so maybe that's really what's going on here. Perhaps my confusion would been avoided with a more detailed error message such as:

```
Cannot process request for application %d's boxes with max=%d. There are more boxes for application %d than are afforded by MaxAPIBoxPerApplication limit %d.
```

Reply:
> I guess this is a way to tell that there are more boxes to see than the server can make available

Correct.  I'd tweak the comment to say "than the server _is configured to_ make available". 

>  Perhaps my confusion would been avoided

* I don't disagree, but I prefer to keep as is.  If you feel strongly, let's move to a verbal.
* The messaging looks to me to follow the convention in `AccountInformation`.  I am reluctant to either deviate from existing conventions or open discussion to suggest changing both.

Reply:
Had a verbal with @tzaffi and agreed to keep as is + add an inline comment explaining _why_ `maxBoxThreshold + 1` happens.

Reply:
Resolving - https://github.com/algorand/go-algorand/pull/4543/commits/54385ac6fefb6f3409b4e6d5b3d9de2d1dd422c2 adds an inline comment.